### PR TITLE
Add persistent reopenable subagents

### PR DESCRIPTION
**Pull request by sethkarten**
## Summary

Adds persistent, reopenable subagents. Today RLM subagents are ephemeral: every `rlm(...)` call mints a fresh `sub-<uuid>` directory and session, so once a run finishes its conversation can be inspected but never continued. This makes persistence opt-in per call so a subagent can be reopened later with its saved history and role.

```python
# First run creates the persistent subagent and stores its role.
r1 = await rlm(
    "Review the auth changes on this branch.",
    persist=True,
    persistent_id="reviewer",
    system_prompt="You are the code reviewer subagent. Be terse and specific.",
)

# Later run with the same id reopens its saved session: system prompt + prior
# history + this new instruction. r2.reopened is True.
r2 = await rlm("Now review the API changes too.", persist=True, persistent_id="reviewer")
```

## Design

- **Opt-in, safe default.** Persistence is off unless `persist=True` (or a `persistent_id`, which implies it). A bare `rlm(prompt)` is unchanged: fresh `sub-<uuid>` dir, no reopen, no sidecar. Unknown kwargs still fail loudly.
- **Reuses the existing session format.** History is the existing append-only session JSONL. The only new artifact is a small `subagent.json` sidecar per subagent (session-file pointer, stored system prompt, run count). No separate incompatible format.
- **Stable identity + location.** A persistent subagent lives under the parent's artifact dir at `persistent-subagents/<slug(id)>/`, so it survives runs and session resumes. Requires a persisted parent session (raises otherwise).
- **Reopen = system prompt + history + new instruction.** On reopen the saved session is loaded and hydrated, the stored system prompt is re-applied as a subagent-scoped append (kept after loader appends), and the new instruction runs on top. An explicit `system_prompt` updates the stored role; omitting it reuses the stored one.
- **All transports.** Inline (non-daemon), in-process `AgentSessionRuntime`, and daemon subagent runtime creators all go through one shared `createSubagentSessionManager` helper that decides create-vs-reopen, so behavior is identical everywhere. A resident finished session for the same id is disposed before reopening so two sessions never share one file.
- **Result surface.** `RLMResult` gains `persistent_id` and `reopened` (Python + TS).
- Agent messaging semantics are untouched (that layer is orthogonal, only for messaging active agents).

## Changed files

- `src/core/persistent-subagents.ts` (new): sidecar store, id slugging, run planning, and the shared subagent SessionManager create/reopen helper.
- `src/core/agent-session.ts`: parse `persist`/`persistent_id`/`system_prompt` kwargs, plan/record persistent runs, hydrate reopened history, apply subagent system prompt, dispose resident session before reopen.
- `src/core/rlm-runtime.ts`, `src/core/agent-session-services.ts`, `src/core/sdk.ts`, `src/main.ts`: thread `subagentSystemPrompt` / `existingSessionFile` and the new result fields through the runtime options.
- `src/core/agent-session-runtime.ts`, `src/modes/daemon/daemon-mode.ts`: reopen through the shared helper and pass the subagent system prompt.
- `src/core/prompts/rlm.ts`: subagent guidance now documents persistent subagents.
- `prime-agent-runtime/src/rlm/__init__.py`: `RLMResult.persistent_id` / `reopened`.
- `docs/kernel-and-rlm-recursion.md`, `CHANGELOG.md`: docs + changelog.
- Tests: `test/persistent-subagents.test.ts` (new) and persistent-subagent cases in `test/agent-session-recursion.test.ts`.

## Validation

- `test/persistent-subagents.test.ts`: 7 passed (store/plan/reopen/hydrate helpers).
- `test/agent-session-recursion.test.ts`: 33 passed, including 6 new cases: create + sidecar, reopen saved history, apply/re-apply system prompt across reopen, non-persistent unchanged, `persist` requires id, non-persisted parent rejected.
- `test/system-prompt.test.ts`: 20 passed (guidance change).
- `prime-agent-runtime` harness tests: 35 passed; Python `RLMResult` payload roundtrip verified.
- `npm run check` (biome + tsgo + installer + browser-smoke): clean.

---

> [!NOTE]
> **Medium Risk**
> Touches RLM child lifecycle, shared session JSONL handoff, and daemon session close paths; mistakes could corrupt transcripts or orphan retained sessions, though the PR adds explicit guards and extensive tests.
> 
> **Overview**
> Adds **opt-in persistent subagents** for `rlm.run` / `rlm(...)`: callers pass `persist=True` (or a `persistent_id`, which implies persistence), optional `system_prompt`, and get a stable on-disk identity under the parent session’s artifact dir (`persistent-subagents/<slug>-<hash>/` plus `subagent.json`). Later runs with the same id **reopen** the saved JSONL history, re-apply the stored role append, and run the new instruction on top; `RLMResult` / `RLMRunResult` expose `persistent_id` and `reopened`. Ephemeral `sub-<uuid>` behavior is unchanged when persistence is not requested.
> 
> New module **`persistent-subagents.ts`** owns sidecar I/O, id slugging, run planning, read-only session probes (so malformed JSONL is never rewritten), and **`createSubagentSessionManager`** so inline, runtime, and daemon paths share the same create-vs-reopen logic.
> 
> **`AgentSession`** parses and validates kwargs (`system_prompt` without persist is rejected), requires a persisted parent, blocks reopen while a retained child is busy or has nested runs, disposes the prior retained session before reopen (with bash drain / `beginReopenTeardown` so late writes cannot corrupt the shared file), restores the child’s last model/thinking from the saved session, and tallies usage/turns/answer from **this run’s** `message_end` events only. **`waitForEventQueueIdle`** closes a race after `waitForIdle`.
> 
> **Daemon** adds `closeRetainedRlmSubagentRuntime`, a **`reopened`** close reason (no archive, no `session_closed` broadcast for the handoff), and subtree-scoped stamping of `activeSessionId` on nested `rlm_child_update` events. Docs, RLM guidance, changelog, and broad tests cover store, reopen, and teardown edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 52af077759706d57473b6cb5b574085a0675e1fd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add persistent reopenable subagents with session history, model, and role restoration
> - Introduces `persist`, `persistent_id`, and `system_prompt` kwargs to `rlm.run` that opt a subagent into persistent mode, storing session files and sidecar metadata under a stable on-disk directory keyed by `persistent_id`.
> - When a persistent subagent is invoked again, it reopens the prior session file, restoring conversation history, model, thinking level, and system prompt rather than starting fresh.
> - Adds `persistent_id` and `reopened` fields to `RLMResult` (Python) and the `rlm.run` result object (TypeScript) so callers can detect whether prior history was resumed.
> - In daemon mode, a session closed for reopen skips archiving and suppresses `session_closed` broadcasts; any prior retained child session with the same `rlmChildId` is fully torn down (including bash drain) before the new session opens.
> - Fixes `rlm_child_update` stamping to scope child lookups to the correct ancestry subtree, avoiding mismatches with unrelated sessions.
> - Risk: reopen is refused if the retained prior session is still busy (streaming, compacting, or running bash) or has active nested subagents; callers must handle this rejection.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 52af077.</sup>
> 
>

### CODE REVIEW

File: packages/coding-agent/src/core/agent-session.ts [hunk 1]
```diff
@@ -5421,12 +5453,20 @@ export class AgentSession {
...
+		// and thinking level so it can be restored on later reopen.
+		const reopened = childSessionManager.buildSessionContext().messages.length > 0;
+		if (!reopened) {
+			childSessionManager.appendModelChange(options.model.provider, options.model.id);
+			childSessionManager.appendThinkingLevelChange(options.thinkingLevel);
 		}
-		childSessionManager.appendModelChange(options.model.provider, options.model.id);
-		childSessionManager.appendThinkingLevelChange(options.thinkingLevel);
 
 		const childAgent = new Agent({
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:5802]: Fixed in 534c90cf. A reopened persistent subagent now restores its last-used model and thinking level from its saved session via `_resolveReopenedSubagentSettings`, which reads the session's `model_change`/`thinking_level_change` (through `buildSessionContext`) and passes them into the runtime options for all transports, falling back to the parent's model only when the saved provider auth is no longer configured.

File: packages/coding-agent/src/core/agent-session.ts [hunk 2]
```diff
@@ -5499,6 +5543,24 @@ export class AgentSession {
 		return this._activeRlmChildRuns.get(childId)?.status;
 	}
 
+	/**
+	 * Dispose a retained (finished, resident) child session by id and drop its event
+	 * forwarder. Used when a persistent subagent is reopened under the same node id, so
+	 * the prior resident session releases the session file before the new one opens it.
+	 */
+	private _disposeRetainedRlmChildSession(childId: string): void {
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts]: Fixed in 534c90cf. `_disposeRetainedRlmChildSession` now detaches the retained session from the maps synchronously and returns the `disposeAsync()` promise; `_startRlmChildRun` awaits it inside the run task before creating the new runtime, so the prior child's kernel finishes flushing its snapshot before the reused session file is reopened.

File: packages/coding-agent/src/core/persistent-subagents.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	/** System prompt (append) stored with the subagent and re-applied on reopen. */
+	systemPrompt?: string;
+	createdAt: string;
+	updatedAt: string;
+	/** Number of times this subagent has been run (created + reopened). */
+	runCount: number;
+}
+
+/** Slug a persistent-subagent id into a filesystem-safe directory name. */
+export function slugifyPersistentSubagentId(id: string): string {
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/persistent-subagents.ts:54]: Fixed in 534c90cf. `slugifyPersistentSubagentId` now appends a short sha256 hash of the exact id to the readable prefix, so ids that normalize identically (case/punctuation/spacing/length) resolve to distinct directories and node ids instead of sharing one.

File: packages/coding-agent/src/core/agent-session.ts [hunk 4]
```diff
@@ -5580,11 +5642,49 @@ export class AgentSession {
...
+		const rawSystemPrompt = kwargs.system_prompt;
+		if (rawSystemPrompt !== undefined && typeof rawSystemPrompt !== "string") {
+			throw new Error("rlm.run system_prompt must be a string");
+		}
+
+		const persistentId = rawPersistentId?.trim() ? rawPersistentId.trim() : undefined;
+		const persist = rawPersist === true || persistentId !== undefined;
+		if (persist && !persistentId) {
+			throw new Error("rlm.run persist requires a persistent_id so the subagent can be reopened later");
+		}
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:6094]: Fixed in 534c90cf. `_parsePersistentSubagentKwargs` now rejects `system_prompt` when persistence is not requested (`system_prompt requires persist=true and a persistent_id`) instead of silently dropping it.

File: packages/coding-agent/src/core/persistent-subagents.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+ * Deterministic node id for a persistent subagent, derived from its id. Stable
+ * across reopens so the parent's child-run map and inspector reference one node.
+ */
+export function persistentSubagentNodeId(id: string): string {
+	return `persist-${slugifyPersistentSubagentId(id)}`;
+}
+
+/** Directory that holds a persistent subagent's session file and sidecar. */
+export function persistentSubagentDir(rootDir: string, id: string): string {
+	return join(rootDir, PERSISTENT_SUBAGENTS_DIR_NAME, slugifyPersistentSubagentId(id));
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/persistent-subagents.ts:76]: Fixed in 534c90cf. Distinct `persistent_id` values no longer share storage: `slugifyPersistentSubagentId` appends a short sha256 hash of the exact id, so ids that normalize identically map to different directories and sidecars.

File: packages/coding-agent/src/core/persistent-subagents.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+		createdAt: record?.createdAt ?? now,
+		updatedAt: now,
+		runCount: (record?.runCount ?? 0) + 1,
+	};
+	return {
+		id: options.id,
+		nodeId: persistentSubagentNodeId(options.id),
+		dir,
+		existingSessionFile,
+		reopened: existingSessionFile !== undefined,
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/persistent-subagents.ts:247]: Fixed in 534c90cf. `reopened` is now gated on `persistentSubagentSessionHasHistory` (the saved session actually has conversation turns), so a metadata-only JSONL no longer reports `reopened: true` while starting fresh and duplicating model/thinking entries; the flag and the hydration decision are resolved in one place in `planPersistentSubagentRun`.

File: packages/coding-agent/src/core/agent-session.ts [hunk 7]
```diff
@@ -5499,6 +5573,28 @@ export class AgentSession {
...
+		if (unsubscribe) {
+			unsubscribe();
+			this._retainedRlmChildUnsubscribes.delete(childId);
+		}
+		const retained = this._retainedRlmChildSessions.get(childId);
+		if (!retained) {
+			return Promise.resolve();
+		}
+		this._retainedRlmChildSessions.delete(childId);
+		return retained.disposeAsync().catch(() => undefined);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts]: Fixed in ac90b52f. Reopening now evicts the prior finished subagent from the host registry, not just the AgentSession: `SubagentRuntimeHost` gained `closeRetainedRlmSubagentRuntime`, and the daemon implements it by locating the resident child `ActiveSessionState` (by `parentActiveSessionId` + `rlmChildId`) and closing it before the new run registers. Core routes retained-session disposal through the host when it is present, so observe/message/status can no longer target the stale disposed session and a duplicate child id can't be registered.

File: packages/coding-agent/src/core/persistent-subagents.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+	return resolve(candidates[candidates.length - 1]);
+}
+
+/**
+ * Whether a session file holds actual conversation turns (not just metadata like
+ * model/thinking-level entries). Used to decide whether reopening genuinely
+ * continues history: a metadata-only or unreadable file is treated as no history,
+ * so the run starts fresh instead of reporting a reopen that hydrates nothing.
+ */
+export function persistentSubagentSessionHasHistory(sessionFile: string): boolean {
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/persistent-subagents.ts:187]: Fixed in ac90b52f. The probe no longer goes through `SessionManager.open` (which rewrites a malformed session on empty parse). `persistentSubagentSessionHasHistory` and the model/thinking-level restore now use a read-only `readPersistentSubagentSessionContext` built on `loadEntriesFromFile` + `buildSessionContext`, which never writes. Added a regression test asserting a malformed `.jsonl` is left byte-for-byte unchanged after probing.

File: packages/coding-agent/src/core/agent-session.ts [hunk 9]
```diff
@@ -5725,11 +5923,23 @@ export class AgentSession {
...
+				// Record the persistent subagent's sidecar (session file + system prompt) so a
+				// later run with the same id reopens this exact history.
+				if (persistentPlan) {
+					const sessionFile = child.sessionFile;
+					savePersistentSubagentRecord(persistentPlan.dir, {
+						...persistentPlan.record,
+						sessionFile: sessionFile ?? persistentPlan.record.sessionFile,
+						updatedAt: new Date().toISOString(),
+					});
+				}
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:6366]: Fixed in 7f9d7d66. The `subagent.json` sidecar (id, stored `system_prompt`, run count) is now written up front in `_startRlmChildRun` when a persistent run starts, before the run can fail. A cancelled/errored first run therefore still records the role, so a later reopen keeps it instead of recovering only the leftover session JSONL. The success path still refreshes the sidecar with the final session file. Added a regression test that a failed first run still persists the system prompt.

File: packages/coding-agent/src/core/agent-session.ts [hunk 10]
```diff
@@ -5725,11 +5929,23 @@ export class AgentSession {
...
 				const result: RlmRunResult = {
 					answer,
 					usage,
 					turns: child._assistantTurnCount(),
 					session_dir: childSessionDir,
+					persistent_id: persistentPlan?.id,
+					reopened: persistentPlan?.reopened,
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:6373]: Fixed in f81666c1. The run now snapshots the child's assistant messages at creation (`_snapshotAssistantMessages`) and excludes them from `_usageForCurrentMessages`, `_assistantUsageForCurrentMessages`, and `_assistantTurnCount`, so a reopened persistent subagent reports only the current run's usage/turns and attributes only that run's usage to the parent. Added regression tests asserting a reopened run reports single-run usage/turns and adds only its own usage to the parent aggregate.

File: packages/coding-agent/src/core/agent-session.ts
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:1]: Fixed in 74ed36a3. Before disposing the retained session on reopen, `_startRlmChildRun` now checks `retained.hasRunningRlmChildren()` and rejects the reopen with a clear error (`... has nested subagents still running; wait for them to finish before reopening`) instead of silently tearing down in-flight nested subagents. Added a regression test asserting reopen is rejected while the retained session reports running nested subagents.

File: packages/coding-agent/src/core/agent-session.ts [hunk 11]
```diff
@@ -5821,8 +6055,8 @@ export class AgentSession {
 					throw new Error(run.error ?? "RLM child cancelled");
 				}
 				const answer = child.getLastAssistantText() ?? "";
-				const usage = child._usageForCurrentMessages();
-				const assistantUsage = child._assistantUsageForCurrentMessages();
+				const usage = child._usageForCurrentMessages(priorAssistantMessages);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts]: Fixed in e02552f2. `getLastAssistantText` now accepts the same `priorAssistantMessages` exclusion set used for usage/turns, so on reopen the answer reflects only the current run. A run that produces no new usable assistant text returns an empty answer instead of leaking the previous run's answer while reporting zero turns. Added a regression test covering an empty-text reopen.

File: packages/coding-agent/src/core/agent-session.ts [hunk 12]
```diff
@@ -5687,11 +5818,52 @@ export class AgentSession {
...
+			throw new Error("rlm.run persistent_id must be a string");
+		}
+		const rawSystemPrompt = kwargs.system_prompt;
+		if (rawSystemPrompt !== undefined && typeof rawSystemPrompt !== "string") {
+			throw new Error("rlm.run system_prompt must be a string");
+		}
+
+		const persistentId = rawPersistentId?.trim() ? rawPersistentId.trim() : undefined;
+		const systemPrompt = rawSystemPrompt?.trim() ? rawSystemPrompt.trim() : undefined;
+		const persist = rawPersist === true || persistentId !== undefined;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts]: Fixed in e02552f2. `persist` is no longer forced true by a `persistent_id` when `persist` is explicitly false: `rawPersist === true || (rawPersist !== false && persistentId !== undefined)`. Because `persist=false` with `persistent_id`/`system_prompt` is contradictory, it is now rejected with a clear error rather than silently persisting. Added tests for the rejection and for `persist=false` running ephemerally.

File: packages/coding-agent/src/core/agent-session.ts [hunk 13]
```diff
@@ -5764,10 +5996,20 @@ export class AgentSession {
 				if (isRlmChildRunCancelled(run)) {
 					throw new Error(run.error ?? "RLM child cancelled");
 				}
+				// Wait for a reopened persistent subagent's prior resident session to finish
+				// tearing down (its kernel flushes a final snapshot) before opening the reused
+				// session file, so the new run can't restore from a half-written snapshot.
+				if (retainedDispose) {
+					await retainedDispose;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:6246]: Fixed in 89abe481. Added an `isRlmChildRunCancelled(run)` check immediately after `await retainedDispose`, before `_createRlmSubagentRuntime`. A cancel that lands during the dispose window (while `run.abort` is still the no-op) now aborts the reopen before opening the session or starting a new runtime. Added a regression test that cancels the reopen while the retained session's dispose is blocked and asserts the new runtime never starts.

File: packages/coding-agent/src/core/agent-session.ts [hunk 14]
```diff
@@ -5687,11 +5818,60 @@ export class AgentSession {
...
+		const rawPersistentId = kwargs.persistent_id;
+		if (rawPersistentId !== undefined && typeof rawPersistentId !== "string") {
+			throw new Error("rlm.run persistent_id must be a string");
+		}
+		const rawSystemPrompt = kwargs.system_prompt;
+		if (rawSystemPrompt !== undefined && typeof rawSystemPrompt !== "string") {
+			throw new Error("rlm.run system_prompt must be a string");
+		}
+
+		const persistentId = rawPersistentId?.trim() ? rawPersistentId.trim() : undefined;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:6081]: Fixed in 89abe481. `_parsePersistentSubagentKwargs` now rejects a supplied-but-blank `persistent_id` (`rlm.run persistent_id must not be blank`) instead of folding it to undefined and silently running ephemerally. Added a regression test for the blank-id rejection.

File: packages/coding-agent/src/modes/daemon/daemon-mode.ts [hunk 15]
```diff
@@ -964,6 +983,18 @@ export class AgentDaemon {
...
+			// child that is still resident in the daemon registry under the same
+			// rlmChildId; otherwise observe/message/status could target the stale disposed
+			// session, and a new session would register with a duplicate child id.
+			closeRetainedRlmSubagentRuntime: async (childId) => {
+				const childState = this.findSubagentStateByChildId(parentState, childId);
+				if (!childState) {
+					return false;
+				}
+				await this.closeSession(childState, "completed");
+				return true;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/daemon/daemon-mode.ts:1214]: Fixed in c65928a5. Added a `reopened` daemon close reason: `closeRetainedRlmSubagentRuntime` now closes the resident subagent with `reopened`, and `closeSessionOnce` skips the `archived` session_state write (and the empty-draft discard) for that reason, since the next run immediately reopens and continues the same JSONL. Added a test asserting a `reopened` close disposes the runtime and evicts it without appending `{ status: 'archived' }`.

File: packages/coding-agent/src/core/persistent-subagents.ts [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+	}
+	const candidates = readdirSync(dir)
+		.filter((name) => name.endsWith(".jsonl"))
+		.map((name) => join(dir, name));
+	if (candidates.length === 0) {
+		return undefined;
+	}
+	// Newest by name: session ids are uuidv7 (time-ordered), so lexical max is newest.
+	candidates.sort();
+	return resolve(candidates[candidates.length - 1]);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/persistent-subagents.ts]: Fixed in c65928a5. `findPersistentSubagentSessionFile` now unions the sidecar pointer with every `.jsonl` in the directory and returns the newest file that actually has conversation history (falling back to the newest existing file only when none has history yet), so a stale pointer to a metadata-only file can't shadow a newer session with history. `planPersistentSubagentRun` also no longer carries a stale `sessionFile` pointer forward when it starts fresh. Added a regression test that a stale sidecar pointer does not shadow a newer file with history.

File: packages/coding-agent/src/core/agent-session.ts [hunk 17]
```diff
@@ -5407,10 +5427,65 @@ export class AgentSession {
...
+	 */
+	_snapshotAssistantMessages(): ReadonlySet<AgentMessage> {
+		const seen = new Set<AgentMessage>();
 		for (const message of this.agent.state.messages) {
 			if (message.role === "assistant") {
+				seen.add(message);
+			}
+		}
+		return seen;
+	}
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts]: Fixed in efc606ae. The run no longer filters the child's message array by object identity (which breaks when mid-run compaction rebuilds `agent.state.messages`). Instead it accumulates this run's usage, turns, and answer from the child's live `message_end` events: only the current run's turns emit those events, so hydrated prior turns and compaction rebuilds can't inflate usage/turns or leak a prior answer. Accumulation is deduped by message identity, and the empty-answer rule matches `getLastAssistantText`. Removed the now-unused array-scan helpers. Added a regression test that accounting stays single-run across many accumulated prior turns on reopen.

File: packages/coding-agent/src/modes/daemon/daemon-mode.ts [hunk 18]
```diff
@@ -2604,10 +2647,10 @@ export class AgentDaemon {
 				await deleteSessionFile(sessionFile).catch(() => undefined);
 			}
 		}
-		if (persistError && !keepsResumeEntry && reason !== "completed") {
+		if (persistError && !keepsResumeEntry && reason !== "completed" && reason !== "reopened") {
 			throw persistError;
 		}
-		if (cascadeError && !keepsResumeEntry && reason !== "completed") {
+		if (cascadeError && !keepsResumeEntry && reason !== "completed" && reason !== "reopened") {
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/daemon/daemon-mode.ts]: Fixed in 701c681b. `closeSessionOnce` now only spares the session being reopened from archiving; it cascades a normal `completed` close to nested children (`childCloseReason = reason === 'reopened' ? 'completed' : reason`), since those nested children are torn down permanently rather than handed off. Added a test asserting a `reopened` close leaves the reopened session unarchived while its nested child is archived.

File: packages/coding-agent/src/modes/daemon/daemon-mode.ts [hunk 19]
```diff
@@ -2604,10 +2651,10 @@ export class AgentDaemon {
...
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/daemon/daemon-mode.ts]: Fixed in 870a1e65. Removed the `&& reason !== 'reopened'` exception from the `cascadeError` guard, so a failed nested-child teardown propagates on a `reopened` close just like other non-completed reasons. The reopen still intentionally skips only its own archive (via `keepsSessionForReopen`), but a nested child whose dispose throws is no longer silently orphaned. Added a test asserting the nested dispose error propagates.

File: packages/coding-agent/src/core/agent-session.ts [hunk 20]
```diff
@@ -5820,9 +6061,9 @@ export class AgentSession {
 				if (isRlmChildRunCancelled(run)) {
 					throw new Error(run.error ?? "RLM child cancelled");
 				}
-				const answer = child.getLastAssistantText() ?? "";
-				const usage = child._usageForCurrentMessages();
-				const assistantUsage = child._assistantUsageForCurrentMessages();
+				const answer = runAnswer;
+				const usage = runUsage;
+				const assistantUsage = runAssistantUsage;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:6347]: Fixed in a6125e8b. Added `AgentSession.waitForEventQueueIdle()` (awaits the internal `_agentEventQueue` that drives subscribe() listeners) and the run now awaits `child.waitForEventQueueIdle()` right after `child.agent.waitForIdle()`. Since waitForIdle means the agent loop has finished emitting, draining the queue guarantees every `message_end` has been tallied before this run reads usage/turns/answer and does parent usage attribution, closing the race.

File: packages/coding-agent/src/core/agent-session.ts [hunk 21]
```diff
@@ -5606,6 +5668,44 @@ export class AgentSession {
...
+		// _retainedRlmChildSessions but the host still owns the retained runtime.
+		const host = this._subagentRuntimeHost;
+		if (host?.closeRetainedRlmSubagentRuntime) {
+			const closed = await host.closeRetainedRlmSubagentRuntime(childId).catch(() => false);
+			// If the host had nothing to close but the parent still holds the session, fall
+			// back to disposing it directly so it can't linger.
+			if (!closed && retained) {
+				await retained.disposeAsync().catch(() => undefined);
+			}
+			return;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:5920]: Fixed in 1ebf7559. `_disposeRetainedRlmChildSession` no longer wraps `closeRetainedRlmSubagentRuntime` in `.catch(() => false)`; the host teardown error now propagates and fails the reopen (the host returns `false` only when it found nothing to close, and throws on an actual teardown failure). A no-op rejection sink on the synchronously-started promise prevents an unhandled rejection before the task awaits it. Added a test that a throwing host teardown fails the reopen.

File: packages/coding-agent/src/modes/daemon/daemon-mode.ts [hunk 22]
```diff
@@ -2604,9 +2651,12 @@ export class AgentDaemon {
 				await deleteSessionFile(sessionFile).catch(() => undefined);
 			}
 		}
-		if (persistError && !keepsResumeEntry && reason !== "completed") {
+		if (persistError && !keepsResumeEntry && reason !== "completed" && reason !== "reopened") {
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/daemon/daemon-mode.ts:2950]: Fixed in 8103461a. `closeSessionOnce` no longer broadcasts `session_closed` for a `reopened` close, since a reopen is an internal handoff, not a user-visible closure (clients treat every `session_closed` as a connection `closed` error). Clients still detach from the old activeSessionId and re-attach to the reopened subagent session; a nested child closed normally still broadcasts. Added a test asserting the reopened session itself gets no `session_closed` broadcast.

File: packages/coding-agent/src/modes/daemon/daemon-mode.ts [hunk 23]
```diff
@@ -2592,7 +2639,12 @@ export class AgentDaemon {
 		}
 		state.unsubscribe?.();
 		await state.runtime.dispose();
-		this.broadcastToSession(state, { type: "session_closed", activeSessionId: state.activeSessionId, reason });
+		// A reopen is an internal handoff, not a user-visible closure: broadcasting
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/daemon/daemon-mode.ts:2933]: Fixed in be4fc840. Two parts: (1) the `reopened` close now aborts the retained session (bash via `abortBashForClose`, then `session.abort()`) before disposing it, so a live turn can't race the next `SessionManager.open` on the same JSONL. (2) `_startRlmChildRun` now also refuses the reopen when the retained session itself is busy (`isStreaming || isCompacting || isBashRunning`), not only when nested children run — covering a session still streaming/compacting after an agent_message steer. Added a test that reopen is refused while the retained session is streaming.

File: packages/coding-agent/src/core/agent-session.ts [hunk 24]
```diff
@@ -5606,6 +5668,47 @@ export class AgentSession {
...
+		}
+		// Inline mode: dispose the session we hold directly.
+		if (retained) {
+			await retained.disposeAsync().catch(() => undefined);
+		}
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:5934]: Fixed in f9ff8e77. The inline and host-fallback dispose paths now abort the retained session (turn + bash, via `session.abort()`) before `disposeAsync`, matching the daemon `reopened` close, so a late steer that started after the busy check can't still be writing the shared JSONL when the next `SessionManager.open` runs. Extracted a shared `_abortAndDisposeRetainedSession` helper. The in-process `AgentSessionRuntime` host retains the finished session in the parent's map (no separate id-keyed registry entry to evict), so this fallback path is the correct place; the daemon host still handles its own eviction via `closeRetainedRlmSubagentRuntime`. Added a test asserting the inline reopen aborts before disposing.

File: packages/coding-agent/src/core/agent-session.ts [hunk 25]
```diff
@@ -5606,6 +5668,59 @@ export class AgentSession {
...
+	/**
+	 * Abort a retained session's in-flight work (turn + bash) before disposing it, so a
+	 * late steer that started after the reopen busy check can't still be writing the shared
+	 * JSONL when the next SessionManager.open runs. Mirrors the daemon "reopened" close,
+	 * which aborts before dispose; the daemon path handles this in closeSession, so this is
+	 * only for the inline / host-fallback paths that dispose the AgentSession directly.
+	 */
+	private async _abortAndDisposeRetainedSession(retained: AgentSession): Promise<void> {
+		await retained.abort().catch(() => undefined);
+		await retained.disposeAsync().catch(() => undefined);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:5962]: Fixed in 2b3b0074. `_abortAndDisposeRetainedSession` now drains bash before dispose, mirroring the daemon `abortBashForClose`: if `isBashRunning`, it calls `abortBash()` and polls `isBashRunning` (bounded by RETAINED_SESSION_BASH_DRAIN_TIMEOUT_MS=5000ms) until idle, then `abort()` (turn/agent-loop) and `disposeAsync()`. So a late bash started after the reopen busy check can't keep touching the shared session file when the next `SessionManager.open` runs. Added a test that a late-started bash is drained to idle before dispose on inline reopen.

File: packages/coding-agent/src/core/agent-session.ts [hunk 26]
```diff
@@ -5606,6 +5671,69 @@ export class AgentSession {
...
+	}
+
+	/**
+	 * Abort a retained session's in-flight work (turn + bash) before disposing it, so a
+	 * late steer that started after the reopen busy check can't still be writing the shared
+	 * JSONL when the next SessionManager.open runs. Mirrors the daemon "reopened" close,
+	 * which aborts before dispose; the daemon path handles this in closeSession, so this is
+	 * only for the inline / host-fallback paths that dispose the AgentSession directly.
+	 */
+	private async _abortAndDisposeRetainedSession(retained: AgentSession): Promise<void> {
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:5945]: Fixed in 4d98bc03. `recordBashResult` is now a no-op once the session is disposing/disposed, and `_abortAndDisposeRetainedSession` calls `retained.beginReopenTeardown()` (sets the disposing flag) before the bounded bash drain and abort. So even if bash is stuck in `runUserBashLocked()` extension dispatch and the drain times out, a late `recordBashResult` can't append a stale `bashExecution` message to the JSONL the reopened session has taken over. Added a test that a late bash result after `beginReopenTeardown` is not recorded (no session entry, no message).

**Comment by sethkarten** [comment]
Superseded by #390, which implements the agreed parent-scoped retained-subagent lifecycle APIs and naming contract. Closing in favor of #390.
